File: javascript/browse_naturally.js
const { chromium } = require('playwright-extra');
const stealth = require('puppeteer-extra-plugin-stealth')();
const path = require('path');
const fs = require('fs');
const schedule = require('node-schedule');

// Add stealth plugin
chromium.use(stealth);

const COOKIES_FILE = path.join(__dirname, 'twitter_cookies.json');

// Convert browser extension cookies to Playwright format
function convertCookies(rawCookies) {
    return rawCookies.map(cookie => {
        let sameSite = cookie.sameSite || 'Lax';
        
        // Normalize sameSite values
        if (!['Strict', 'Lax', 'None'].includes(sameSite)) {
            if (sameSite === 'no_restriction') {
                sameSite = 'None';
            } else if (sameSite === 'lax') {
                sameSite = 'Lax';
            } else if (sameSite === 'strict') {
                sameSite = 'Strict';
            } else {
                sameSite = 'Lax';
            }
        }

        return {
            name: cookie.name,
            value: cookie.value,
            domain: cookie.domain,
            path: cookie.path || '/',
            expires: cookie.expirationDate || -1,
            httpOnly: cookie.httpOnly || false,
            secure: cookie.secure || false,
            sameSite: sameSite
        };
    });
}

// Random sleep to simulate human behavior
function randomSleep(minSeconds = 1, maxSeconds = 5) {
    const ms = (Math.random() * (maxSeconds - minSeconds) + minSeconds) * 1000;
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Human-like scrolling
async function humanLikeScroll(page) {
    const scrollAmount = Math.floor(Math.random() * 500) + 300;
    await page.evaluate((amount) => {
        window.scrollBy(0, amount);
    }, scrollAmount);
    await randomSleep(1, 3);
}

// Natural browsing behavior
async function browseTwitterNaturally() {
    if (!fs.existsSync(COOKIES_FILE)) {
        console.error(`Error: ${COOKIES_FILE} not found!`);
        console.log('\nHow to export cookies:');
        console.log('1. Install "Cookie-Editor" extension in your browser');
        console.log('2. Go to twitter.com and make sure you\'re logged in');
        console.log('3. Click the Cookie-Editor extension icon');
        console.log('4. Click "Export" -> "Export as JSON"');
        console.log(`5. Save the file as "${COOKIES_FILE}" in this directory`);
        return;
    }

    console.log(`Loading cookies from ${COOKIES_FILE}...`);
    const rawCookies = JSON.parse(fs.readFileSync(COOKIES_FILE, 'utf8'));
    const cookies = convertCookies(rawCookies);

    // Random browsing duration between 2-10 minutes
    const browseDuration = Math.floor(Math.random() * 480 + 120); // 120-600 seconds
    const minutes = Math.floor(browseDuration / 60);
    const seconds = browseDuration % 60;
    
    console.log('\n' + '='.repeat(60));
    console.log('🤖 Starting natural browsing session');
    console.log(`📅 Time: ${new Date().toLocaleString()}`);
    console.log(`⏱️  Duration: ${minutes} minutes ${seconds} seconds`);
    console.log('='.repeat(60) + '\n');

    const browser = await chromium.launch({
        headless: false,
        args: [
            '--disable-blink-features=AutomationControlled',
            '--no-sandbox',
            '--disable-setuid-sandbox'
        ]
    });

    const context = await browser.newContext({
        viewport: { width: 1280, height: 720 },
        userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
        locale: 'en-US',
        timezoneId: 'America/New_York'
    });

    // Add cookies to context
    await context.addCookies(cookies);
    
    const page = await context.newPage();
    const startTime = Date.now();
    const endTime = startTime + (browseDuration * 1000);

    try {
        console.log('🌐 Opening Twitter...');
        await page.goto('https://twitter.com/home');
        await randomSleep(2, 4);

        // Close any popups
        try {
            await page.getByTestId('app-bar-close').click({ timeout: 2000 });
        } catch (e) {
            // Ignore if popup doesn't exist
        }

        let actionCount = 0;

        while (Date.now() < endTime) {
            const remaining = Math.floor((endTime - Date.now()) / 1000);
            if (remaining <= 0) break;

            // Randomly choose an action
            const actions = ['scroll', 'read', 'explore', 'profile', 'trending'];
            const action = actions[Math.floor(Math.random() * actions.length)];

            actionCount++;
            console.log(`\n[${actionCount}] ⏱️  ${remaining}s remaining - Action: ${action}`);

            if (action === 'scroll') {
                // Scroll through timeline
                const scrollCount = Math.floor(Math.random() * 3) + 2;
                console.log(`  📜 Scrolling ${scrollCount} times...`);
                
                for (let i = 0; i < scrollCount; i++) {
                    await humanLikeScroll(page);
                    
                    // Occasionally stop to "read" a tweet
                    if (Math.random() < 0.3) {
                        console.log('  👀 Reading a tweet...');
                        await randomSleep(3, 8);
                    }
                }
            } else if (action === 'read') {
                // Stop and read the current view
                console.log('  📖 Reading tweets...');
                await randomSleep(5, 15);
            } else if (action === 'explore') {
                // Navigate to Explore page
                try {
                    console.log('  🔍 Checking Explore page...');
                    await page.goto('https://twitter.com/explore');
                    await randomSleep(2, 4);

                    // Scroll a bit
                    const scrolls = Math.floor(Math.random() * 2) + 1;
                    for (let i = 0; i < scrolls; i++) {
                        await humanLikeScroll(page);
                    }

                    await randomSleep(3, 7);

                    // Go back to home
                    await page.goto('https://twitter.com/home');
                    await randomSleep(2, 3);
                } catch (e) {
                    console.log(`  ⚠️  Could not access Explore: ${e.message}`);
                }
            } else if (action === 'profile') {
                // Check own profile occasionally
                try {
                    console.log('  👤 Checking profile...');
                    await page.goto('https://twitter.com/home');
                    await randomSleep(2, 4);
                    await humanLikeScroll(page);
                    await randomSleep(3, 6);
                } catch (e) {
                    console.log(`  ⚠️  Could not access profile: ${e.message}`);
                }
            } else if (action === 'trending') {
                // Look at trending topics
                try {
                    console.log('  🔥 Viewing trending topics...');
                    await page.goto('https://twitter.com/explore/tabs/trending');
                    await randomSleep(3, 6);

                    // Scroll through trending
                    const scrolls = Math.floor(Math.random() * 2) + 1;
                    for (let i = 0; i < scrolls; i++) {
                        await humanLikeScroll(page);
                    }

                    // Go back to home
                    await page.goto('https://twitter.com/home');
                    await randomSleep(2, 3);
                } catch (e) {
                    console.log(`  ⚠️  Could not access trending: ${e.message}`);
                }
            }

            // Random pause between actions
            await randomSleep(2, 5);
        }
    } catch (error) {
        console.log(`\n❌ Error during browsing: ${error.message}`);
    } finally {
        const actualDuration = Math.floor((Date.now() - startTime) / 1000);
        const finalMinutes = Math.floor(actualDuration / 60);
        const finalSeconds = actualDuration % 60;

        console.log('\n' + '='.repeat(60));
        console.log('✅ Browsing session completed!');
        console.log(`⏱️  Total time: ${finalMinutes} minutes ${finalSeconds} seconds`);
        console.log(`🎯 Actions performed: ${actionCount}`);
        console.log('='.repeat(60) + '\n');

        await randomSleep(2, 4);
        await context.close();
        await browser.close();
    }
}

// Schedule random browsing sessions
function scheduleRandomBrowsing() {
    console.log('\n🤖 Twitter Natural Browsing Scheduler');
    console.log('='.repeat(60));
    console.log('This will schedule random browsing sessions throughout the day');
    console.log('to make your account appear more human-like to X/Twitter.\n');

    // Schedule 3-5 random times per day
    const sessionsPerDay = Math.floor(Math.random() * 3) + 3;

    console.log(`📅 Scheduling ${sessionsPerDay} browsing sessions today...`);

    const scheduledTimes = [];
    for (let i = 0; i < sessionsPerDay; i++) {
        // Random hour between 8 AM and 11 PM
        const hour = Math.floor(Math.random() * 16) + 8;
        const minute = Math.floor(Math.random() * 60);

        const timeStr = `${hour.toString().padStart(2, '0')}:${minute.toString().padStart(2, '0')}`;
        scheduledTimes.push(timeStr);

        // Schedule the job
        const rule = new schedule.RecurrenceRule();
        rule.hour = hour;
        rule.minute = minute;

        schedule.scheduleJob(rule, async () => {
            console.log(`\n⏰ Scheduled browsing session starting at ${new Date().toLocaleTimeString()}...`);
            await browseTwitterNaturally();
        });

        console.log(`  ⏰ Session ${i + 1}: ${timeStr}`);
    }

    console.log('\n✅ Scheduled! Waiting for next session...');
    console.log('💡 Press Ctrl+C to stop the scheduler\n');

    // Keep the script running
    process.on('SIGINT', () => {
        console.log('\n\n👋 Scheduler stopped by user');
        process.exit(0);
    });
}

// Main function
async function main() {
    const command = process.argv[2];

    if (!command) {
        console.log('\nUsage:');
        console.log('  node browse_naturally.js browse    - Browse once now');
        console.log('  node browse_naturally.js schedule  - Schedule random browsing sessions');
        return;
    }

    if (command === 'browse') {
        await browseTwitterNaturally();
    } else if (command === 'schedule') {
        scheduleRandomBrowsing();
    } else {
        console.log(`Unknown command: ${command}`);
        console.log('\nAvailable commands:');
        console.log('  browse    - Browse once now');
        console.log('  schedule  - Schedule random browsing sessions');
    }
}

main();

File: python/browse_naturally.py
import os
import random
import time
import json
from datetime import datetime, timedelta
from playwright.sync_api import Playwright, sync_playwright
import schedule

COOKIES_FILE = "twitter_cookies.json"


def random_sleep(min_seconds=1, max_seconds=5):
    """Sleep for a random amount of time to simulate human behavior"""
    time.sleep(random.uniform(min_seconds, max_seconds))


def human_like_scroll(page):
    """Simulate human-like scrolling behavior"""
    scroll_amount = random.randint(300, 800)
    page.evaluate(f"window.scrollBy(0, {scroll_amount})")
    random_sleep(1, 3)


def browse_twitter_naturally(playwright: Playwright) -> None:
    """Browse Twitter like a real human for random duration"""
    if not os.path.exists(COOKIES_FILE):
        print(f"Error: {COOKIES_FILE} not found!")
        print("\nHow to export cookies:")
        print("1. Install 'Cookie-Editor' extension in your browser")
        print("2. Go to twitter.com and make sure you're logged in")
        print("3. Click the Cookie-Editor extension icon")
        print("4. Click 'Export' -> 'Export as JSON'")
        print(f"5. Save the file as '{COOKIES_FILE}' in this directory")
        return

    print(f"Loading cookies from {COOKIES_FILE}...")
    with open(COOKIES_FILE, 'r') as f:
        cookies_raw = json.load(f)
    
    # Convert cookies to Playwright format
    cookies = []
    for cookie in cookies_raw:
        # Handle sameSite attribute
        same_site = cookie.get('sameSite', 'Lax')
        if same_site not in ['Strict', 'Lax', 'None']:
            if same_site == 'no_restriction':
                same_site = 'None'
            elif same_site.lower() == 'lax':
                same_site = 'Lax'
            elif same_site.lower() == 'strict':
                same_site = 'Strict'
            else:
                same_site = 'Lax'
        
        cookies.append({
            'name': cookie['name'],
            'value': cookie['value'],
            'domain': cookie['domain'],
            'path': cookie.get('path', '/'),
            'expires': cookie.get('expirationDate', -1),
            'httpOnly': cookie.get('httpOnly', False),
            'secure': cookie.get('secure', False),
            'sameSite': same_site
        })

    # Random browsing duration between 2-10 minutes
    browse_duration = random.randint(120, 600)  # seconds
    print(f"\n{'='*60}")
    print(f"🤖 Starting natural browsing session")
    print(f"📅 Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print(f"⏱️  Duration: {browse_duration // 60} minutes {browse_duration % 60} seconds")
    print(f"{'='*60}\n")

    browser = playwright.chromium.launch(headless=False)
    context = browser.new_context(
        viewport={'width': 1280, 'height': 720},
        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
        locale='en-US',
        timezone_id='America/New_York'
    )
    
    # Add cookies to context
    context.add_cookies(cookies)
    
    page = context.new_page()

    start_time = time.time()
    end_time = start_time + browse_duration

    try:
        print("🌐 Opening Twitter...")
        page.goto("https://twitter.com/home")
        random_sleep(2, 4)
        
        # Close any popups
        try:
            page.get_by_test_id("app-bar-close").click(timeout=2000)
        except:
            pass

        action_count = 0
        
        while time.time() < end_time:
            remaining = int(end_time - time.time())
            if remaining <= 0:
                break
                
            # Randomly choose an action
            actions = ['scroll', 'read', 'explore', 'profile', 'trending']
            action = random.choice(actions)
            
            action_count += 1
            print(f"\n[{action_count}] ⏱️  {remaining}s remaining - Action: {action}")
            
            if action == 'scroll':
                # Scroll through timeline
                scroll_count = random.randint(2, 5)
                print(f"  📜 Scrolling {scroll_count} times...")
                for i in range(scroll_count):
                    human_like_scroll(page)
                    
                    # Occasionally stop to "read" a tweet
                    if random.random() < 0.3:
                        print(f"  👀 Reading a tweet...")
                        random_sleep(3, 8)
            
            elif action == 'read':
                # Stop and read the current view
                print(f"  📖 Reading tweets...")
                random_sleep(5, 15)
            
            elif action == 'explore':
                # Navigate to Explore page
                try:
                    print(f"  🔍 Checking Explore page...")
                    page.goto("https://twitter.com/explore")
                    random_sleep(2, 4)
                    
                    # Scroll a bit
                    for _ in range(random.randint(1, 3)):
                        human_like_scroll(page)
                    
                    random_sleep(3, 7)
                    
                    # Go back to home
                    page.goto("https://twitter.com/home")
                    random_sleep(2, 3)
                except Exception as e:
                    print(f"  ⚠️  Could not access Explore: {e}")
            
            elif action == 'profile':
                # Check own profile occasionally
                try:
                    print(f"  👤 Checking profile...")
                    # Click on profile icon
                    page.goto("https://twitter.com/home")
                    random_sleep(2, 4)
                    human_like_scroll(page)
                    random_sleep(3, 6)
                except Exception as e:
                    print(f"  ⚠️  Could not access profile: {e}")
            
            elif action == 'trending':
                # Look at trending topics
                try:
                    print(f"  🔥 Viewing trending topics...")
                    page.goto("https://twitter.com/explore/tabs/trending")
                    random_sleep(3, 6)
                    
                    # Scroll through trending
                    for _ in range(random.randint(1, 3)):
                        human_like_scroll(page)
                    
                    # Go back to home
                    page.goto("https://twitter.com/home")
                    random_sleep(2, 3)
                except Exception as e:
                    print(f"  ⚠️  Could not access trending: {e}")
            
            # Random pause between actions
            random_sleep(2, 5)

    except Exception as e:
        print(f"\n❌ Error during browsing: {e}")
    
    finally:
        actual_duration = int(time.time() - start_time)
        print(f"\n{'='*60}")
        print(f"✅ Browsing session completed!")
        print(f"⏱️  Total time: {actual_duration // 60} minutes {actual_duration % 60} seconds")
        print(f"🎯 Actions performed: {action_count}")
        print(f"{'='*60}\n")
        
        random_sleep(2, 4)
        context.close()
        browser.close()


def schedule_random_browsing():
    """Schedule browsing sessions at random times throughout the day"""
    print("\n🤖 Twitter Natural Browsing Scheduler")
    print("="*60)
    print("This will schedule random browsing sessions throughout the day")
    print("to make your account appear more human-like to X/Twitter.\n")
    
    # Schedule 3-5 random times per day
    sessions_per_day = random.randint(3, 5)
    
    print(f"📅 Scheduling {sessions_per_day} browsing sessions today...")
    
    # Generate random times for today
    scheduled_times = []
    for i in range(sessions_per_day):
        # Random hour between 8 AM and 11 PM
        hour = random.randint(8, 23)
        minute = random.randint(0, 59)
        time_str = f"{hour:02d}:{minute:02d}"
        scheduled_times.append(time_str)
        
        schedule.every().day.at(time_str).do(lambda: browse_with_playwright())
        print(f"  ⏰ Session {i+1}: {time_str}")
    
    print(f"\n✅ Scheduled! Waiting for next session...")
    print(f"💡 Press Ctrl+C to stop the scheduler\n")
    
    try:
        while True:
            schedule.run_pending()
            time.sleep(60)  # Check every minute
    except KeyboardInterrupt:
        print("\n\n👋 Scheduler stopped by user")


def browse_with_playwright():
    """Wrapper to run browsing with playwright"""
    with sync_playwright() as playwright:
        browse_twitter_naturally(playwright)


def main():
    import sys
    
    if len(sys.argv) < 2:
        print("\nUsage:")
        print("  python browse_naturally.py browse    - Browse once now")
        print("  python browse_naturally.py schedule  - Schedule random browsing sessions")
        return
    
    command = sys.argv[1]
    
    if command == "browse":
        with sync_playwright() as playwright:
            browse_twitter_naturally(playwright)
    
    elif command == "schedule":
        schedule_random_browsing()
    
    else:
        print(f"Unknown command: {command}")
        print("\nAvailable commands:")
        print("  browse    - Browse once now")
        print("  schedule  - Schedule random browsing sessions")


if __name__ == "__main__":
    main()
